File: blukutukhttp/src/main/java/com/robinyonathan/blukutukhttp/ProgressRequestBody.java
/*
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Robin Yonathan <robin@robinyonathan.com>, 2017.
 */

/*
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Robin Yonathan <robin@robinyonathan.com>, 2017.
 */

package com.robinyonathan.blukutukhttp;


import androidx.annotation.NonNull;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;
import okio.Sink;

/**
 * Created by Robin Yonathan on 14/08/17.
 */

public class ProgressRequestBody extends RequestBody {

    private RequestBody mDelegate;
    private Listener mListener;

    ProgressRequestBody(RequestBody delegate, Listener listener) {
        mDelegate = delegate;
        mListener = listener;
    }

    @Override
    public MediaType contentType() {
        return mDelegate.contentType();
    }

    @Override
    public long contentLength() {
        try {
            return mDelegate.contentLength();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    @Override
    public void writeTo(@NonNull BufferedSink sink) throws IOException {
        CountingSink mCountingSink = new CountingSink(sink);
        BufferedSink bufferedSink = Okio.buffer(mCountingSink);
        mDelegate.writeTo(bufferedSink);
        bufferedSink.flush();
    }

    protected final class CountingSink extends ForwardingSink {
        private long bytesWritten = 0;

        CountingSink(Sink delegate) {
            super(delegate);
        }

        @Override
        public void write(@NonNull Buffer source, long byteCount) throws IOException {
            super.write(source, byteCount);
            bytesWritten += byteCount;
            mListener.onProgress((int) (100F * bytesWritten / contentLength()));
        }
    }

    public interface Listener {
        void onProgress(int progress);
    }
}

File: blukutukhttp/src/main/java/com/robinyonathan/blukutukhttp/ProgressRequestBody.kt
/*
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Robin Yonathan <robin@robinyonathan.com>, 2017.
 */

/*
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * Written by Robin Yonathan <robin@robinyonathan.com>, 2017.
 */

package com.robinyonathan.blukutukhttp


import okhttp3.MediaType
import okhttp3.RequestBody
import okio.*
import java.io.IOException

/**
 * Created by Robin Yonathan on 14/08/17.
 */

class ProgressRequestBody internal constructor(private val mDelegate: RequestBody, private val mListener: Listener) : RequestBody() {

    override fun contentType(): MediaType? {
        return mDelegate.contentType()
    }

    override fun contentLength(): Long {
        try {
            return mDelegate.contentLength()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return -1
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        val mCountingSink = CountingSink(sink)
        val bufferedSink = Okio.buffer(mCountingSink)
        mDelegate.writeTo(bufferedSink)
        bufferedSink.flush()
    }

    private inner class CountingSink internal constructor(delegate: Sink) : ForwardingSink(delegate) {
        private var bytesWritten: Long = 0

        @Throws(IOException::class)
        override fun write(source: Buffer, byteCount: Long) {
            super.write(source, byteCount)
            bytesWritten += byteCount
            mListener.onProgress((100f * bytesWritten / contentLength()).toInt())
        }
    }

    interface Listener {
        fun onProgress(progress: Int)
    }
}
